### DISCUSSION TITLE
Reverse relation doesn’t seem to work - I must be missing something very basic

### DISCUSSION THREAD

**Discussion by nicholasbutlin** [collerek/ormar#552] [Q&A]
Hi, I have been using Ormar and love it. I realise that this is a very, very basic question, but for some reason I cannot get the result I expect. 

Below is an example I made with the same setup as I have in my code and using the example from the tutorials: 

```
import logging
import uuid
from pathlib import Path
from typing import Optional

import databases
import ormar
from databases import DatabaseURL
from pydantic import BaseSettings
from sqlalchemy import MetaData, create_engine

logger = logging.getLogger(__name__)

class DataBaseSettings(BaseSettings):
    DB_USER: str
    DB_PASS: str
    DB_HOST: str
    DB_PORT: str
    DB_NAME: str


class DbManager:

    def __init__(self):
        self.settings = DataBaseSettings()
        self.url = DatabaseURL(
            f'mysql://{self.settings.DB_USER}:{self.settings.DB_PASS}@{self.settings.DB_HOST}:{self.settings.DB_PORT}/{self.settings.DB_NAME}'
        )
        self._url_string = str(self.url.replace(dialect='mysql+pymysql'))
        self.engine = self.getEngine()
        self.metadata = MetaData()
        self.database = databases.Database(self.url)

    def getEngine(self):
        return create_engine(self._url_string,
                             echo=False,
                             connect_args={'connect_timeout': 10})

    def createAll(self):
        self.metadata.create_all(self.engine)

    @property
    def OrmarMeta(self):
        return type('BaseMeta', (ormar.ModelMeta, ), {
            'metadata': self.metadata,
            'database': self.database
        })


db_manager = DbManager()
BaseMeta = db_manager.OrmarMeta


def generate_guid() -> str:
    """
    Utility to generate a GUID.

    Returns:
        str: A GUID string
    """
    return str(uuid.uuid4())


class Department(ormar.Model):

    class Meta(BaseMeta):
        table_name = 'department_table'

    id: int = ormar.UUID(primary_key=True,
                         uuid_format="string",
                         default_factory=generate_guid)
    name: str = ormar.String(max_length=100)


class Course(ormar.Model):

    class Meta(BaseMeta):
        table_name = 'table_course'

    id: int = ormar.UUID(primary_key=True,
                         uuid_format="string",
                         default_factory=generate_guid)
    name: str = ormar.String(max_length=100)
    completed: bool = ormar.Boolean(default=False)
    department: Optional[Department] = ormar.ForeignKey(Department,
                                                        related_name='course')


db_manager.createAll()


async def setup():
    await db_manager.database.connect()
    department = await Department.objects.create(name="Science3”)
    # using a str not the whole model - wondered if this was something to do with it... I’m firing in the dark
    department_id = str(department.id)
    course = await Course.objects.create(name="Math3",
                                         completed=False,
                                         department=department_id)
    dep = await Department.objects.filter(id=department_id).select_all().get()
```

This works fine, and when I look at depot, get the ‘weaker’ and the data populated, dep looks like this: 
```
Department({'id': UUID('98a9bf6c-9fd3-4f1b-b349-c1b009022bce'), 'name': 'Science3', 'course': [Course({'id': UUID('948c2c45-25af-4786-8891-af8cab9b03f4'), 'name': 'Math3', 'completed': False, 'department': <weakproxy at 0x111a5e4a0 to Department at 0x113fa1ee0>})]})
```

And when I check the related names, I get what I would expect - a relation each way 
```
>>> Department._related_names
{'course'}
>>> Course._related_names
{‘department’}
```

When I do the same in my actual code, where there are about 20 interrelated models, I don’t get weak refs, and no data - just the virtual field.
In a couple of case, the virtual side appears not to have been registered at all and there is no reverse relation, just an empty set e.g.: 
I have the equivalent of models Department and Course, but now when I run a similar process to the above I get:
```
Course({ ‘department': Department({'id': UUID('6bdb7eb3-ede1-4e8c-a074-e850bfea3ba5'), 'name': ‘bar’,  ‘courses': None}), 'name': ‘foo’})
>>> Course._related_names
{‘department’}
```

Department:
```
Department({'id': UUID('de6953e2-e191-42a7-856a-b0f136e0167a'), 'name': ‘bar', ‘courses': None})
>>> Department._related_names
set()
```

I have tried: 
- changing table names 
- renaming indexes 
- changing related_names
- recreating the db 

nothing I can do seems to work and I am truly, utterly confused. 

I am using MariaDB 10, python 3.8, 0rmar 0.10.24

Any ideas as to why it just doesn’t work at all would be gratefully received.

**Comment by collerek** [comment]
Hi,

Hard to tell without a minimal reproducible example.
Can you provide a minimal example of your code that does not work (like with 2 models)

Did you check if data is properly created in the database?
The string-ifing uuid can bite you if you select a different string/hex option, otherwise, it should be fine.
Note that to select deeper relations you need to pass `select_all(follow=True)` otherwise only immediately related models of a parent (the one you rub query for) are selected. 
Did you create any pydantic models and have conversions or operate solely on ormar models?

**Reply by nicholasbutlin** [reply]
Hi, Thanks so much for your reply. 

The example above was supposed to be the minimal example of the non working code... but it worked! 
It uses the same basic setup, other than it being a lot simpler and in the same file. 

- The data is created OK
- The UUID seems to fine, because they are found, I can get anything from the DB I like it seems 
- These are just the direct relations - they don’t appear to be being registered or found at all as the weak proxy doesn’t get made 
- Working with ormar models directly. 

I really am totally stumped. Will try to find an example of it not working...

**Reply by nicholasbutlin** [reply]
Although it makes me wonder whether it is some sort of import order issues...

**Reply by nicholasbutlin** [reply]
So working back in a copy of the project, I deleted models and made the structure increasingly simple.
It didn’t work until I defined all the models in the same file. 

Keeping each model in a different module (along with the router and other crud code) and importing them from other modules meant that the refs weren’t created - and no reverse relations. 

When they are all in the same file it worked just fine.  So some sort of import issue. More investigation needed this end...
